Mathlib/Analysis/Analytic/Meromorphic.lean

```diff
@@ -122,4 +131,144 @@ lemma div {f g : 𝕜 → 𝕜} {x : 𝕜} (hf : MeromorphicAt f x) (hg : Meromo
+    rw [AnalyticAt.order_eq_top_iff] at h
+    refine fun ⟨g, hg_an, hg_ne, hg_eq⟩ ↦ hg_ne ?_
+    apply EventuallyEq.eq_of_nhds
+    rw [EventuallyEq, ← AnalyticAt.frequently_eq_iff_eventually_eq hg_an analyticAt_const]
+    apply Eventually.frequently
+    rw [eventually_nhdsWithin_iff] at hg_eq ⊢
+    filter_upwards [h, hg_eq] with z hfz hfz_eq hz
+    rwa [hfz_eq hz, ← mul_smul, smul_eq_zero_iff_right] at hfz
+    exact mul_ne_zero (pow_ne_zero _ (sub_ne_zero.mpr hz)) (zpow_ne_zero _ (sub_ne_zero.mpr hz))
+  · obtain ⟨m, h⟩ := WithTop.ne_top_iff_exists.mp h
+    rw [← h, WithTop.map_coe, ← WithTop.coe_nat, ← WithTop.coe_sub, WithTop.coe_inj]
+    obtain ⟨g, hg_an, hg_ne, hg_eq⟩ := (AnalyticAt.order_eq_nat_iff _ _).mp h.symm
+    replace hg_eq : ∀ᶠ (z : 𝕜) in 𝓝[≠] x, f z = (z - x) ^ (↑m - ↑hf.choose : ℤ) • g z
+    · rw [eventually_nhdsWithin_iff]
+      filter_upwards [hg_eq] with z hg_eq hz
+      rwa [← smul_right_inj <| zpow_ne_zero _ (sub_ne_zero.mpr hz), ← mul_smul,
+      ← zpow_add₀ (sub_ne_zero.mpr hz), ← add_sub_assoc, add_sub_cancel', zpow_ofNat, zpow_ofNat]
+    exact ⟨fun h ↦ ⟨g, hg_an, hg_ne, h ▸ hg_eq⟩,
+      AnalyticAt.unique_eventuallyEq_zpow_smul_nonzero ⟨g, hg_an, hg_ne, hg_eq⟩⟩
+
+/-- Compatibility of notions of `order` for analytic and meromorphic functions. -/
+lemma order_of_analyticAt {f : 𝕜 → E} {x : 𝕜} (hf : AnalyticAt 𝕜 f x) :
+    hf.meromorphicAt.order = hf.order.map (↑·) := by
+  rcases eq_or_ne hf.order ⊤ with ho | ho
+  · rw [ho, WithTop.map_top, order_eq_top_iff]
+    exact (hf.order_eq_top_iff.mp ho).filter_mono nhdsWithin_le_nhds
+  · obtain ⟨n, hn⟩ := WithTop.ne_top_iff_exists.mp ho
+    simp_rw [← hn, WithTop.map_coe, order_eq_int_iff, zpow_ofNat]
+    rcases (hf.order_eq_nat_iff _).mp hn.symm with ⟨g, h1, h2, h3⟩
+    exact ⟨g, h1, h2, h3.filter_mono nhdsWithin_le_nhds⟩
+
+lemma iff_eventuallyEq_zpow_smul_analyticAt {f : 𝕜 → E} {x : 𝕜} : MeromorphicAt f x ↔
+    ∃ (n : ℤ) (g : 𝕜 → E), AnalyticAt 𝕜 g x ∧ ∀ᶠ z in 𝓝[≠] x, f z = (z - x) ^ n • g z := by
+  constructor
+  · intro hf_m
+    rcases eq_or_ne hf_m.order ⊤ with ho | ho
+    · exact ⟨0, 0, analyticAt_const, (hf_m.order_eq_top_iff.mp ho).mp
+        (eventually_of_forall fun z hz ↦ by rw [hz, Pi.zero_apply, smul_zero])⟩
+    · obtain ⟨n, hn⟩ := WithTop.ne_top_iff_exists.mp ho
+      obtain ⟨g, hg_an, _, hg_eq⟩ := (hf_m.order_eq_int_iff _).mp hn.symm
+      exact ⟨n, g, hg_an, hg_eq⟩
+  · rintro ⟨n, g, hg_an, hg_eq⟩
+    refine MeromorphicAt.congr ?_ (EventuallyEq.symm hg_eq)
+    exact (((meromorphicAt_id x).sub (meromorphicAt_const _ x)).zpow _).smul hg_an.meromorphicAt
+
 end MeromorphicAt
+
+/-- Meromorphy of a function on a set. -/
+def MeromorphicOn (f : 𝕜 → E) (U : Set 𝕜) : Prop := ∀ x ∈ U, MeromorphicAt f x
+
+lemma AnalyticOn.meromorphicOn {f : 𝕜 → E} {U : Set 𝕜} (hf : AnalyticOn 𝕜 f U) :
+    MeromorphicOn f U :=
+  fun x hx ↦ (hf x hx).meromorphicAt
+
+lemma meromorphicOn_id {U : Set 𝕜} : MeromorphicOn id U := fun x _ ↦ meromorphicAt_id x
```

Comment:
With the new dot notation in Lean 4, this kind of lemma should be named `MeromorphicOn.id`. The reason is that, if you're proving that a function is `MeromorphicOn`, then Lean 4 will let you use `.id` for `MeromorphicOn.id`, and it turns out to be really convenient. Same thing below for `const`.